File: site/testsuites/testsuite-kotlin/src/test/kotlin/docs/utils/MockOfferingsApiProvider.kt
package website.tbd.developer.site.docs.utils

import tbdex.sdk.httpserver.models.*
import tbdex.sdk.protocol.models.*
import web5.sdk.dids.did.BearerDid

open class MockOfferingsApiProvider: OfferingsApi {

    val dataProvider = MockDataProvider()

    //---------------------------------------------------------------------------//
    // Implementation of interface
    //---------------------------------------------------------------------------//

    override fun getOffering(id: String): Offering {
        val result = dataProvider.get("offering", id ?: "")
        return Offering.parse(result as String)
    }

    override fun getOfferings(): List<Offering> {
        val results = dataProvider.query("offering", "*")
        val offerings = mutableListOf<Offering>()

        for (result in results) {
            val offering = Offering.parse(result as String)
            offerings.add(offering)
        }

        return offerings
    }

    //---------------------------------------------------------------------------//
    // Setup Methods
    //---------------------------------------------------------------------------//

    fun setOffering(id: String, pfiDid: BearerDid) {
        val offering = TestData.getOffering(pfiDid.uri, TestData.getPresentationDefinition())
        offering.sign(pfiDid)
        dataProvider.setupGet("offering", id) {offering }
    }

    fun setOfferings(offeringDids: List<String>) {
        dataProvider.setupGet("offering", "*") {
            offeringDids.map { did -> TestData.getOffering(did, TestData.getPresentationDefinition()) }
        }
    }
}

File: site/testsuites/testsuite-javascript/__tests__/utils/mockOfferingsApiProvider.js
import { MockDataProvider } from './mockDataProvider'
import { Offering } from '@tbdex/protocol'
import { DevTools } from '@tbdex/http-client';

export class MockOfferingsApiProvider {

    constructor() {
        this.dataProvider = new MockDataProvider()
    }

    //---------------------------------------------------------------------------//
    // Implementation of interface
    //---------------------------------------------------------------------------//

    async getOffering(id) {
        this.dataProvider.get('offering', id).then(([result]) => {
            return Offering.create({
                metadata: { from: this.pfiDid },
                data: result.data
            })
        });
    }
  
    getOfferings() {
        this.dataProvider.query('offering', "*").then((results) => {
            const offerings = []
      
            for (let result of results) {
                const offering = Offering.create({
                  metadata: { from: this.pfiDid },
                  data: result.data
                })
                offerings.push(offering)
            }
        
            return offerings
        });
    }

    //---------------------------------------------------------------------------//
    // Setup Methods
    //---------------------------------------------------------------------------//
  
    setOffering(id, offering) {
console.log('called setOffering', id, offering)
        this.dataProvider.setupGet('offering', id, () => {
            return offering
        })
    }

    setOfferings(offeringData) {
        this.dataProvider.setupQuery('offering', "*", () => {
            const offerings = []
            offeringData.forEach(offering => {
                offerings.push(DevTools.createOffering(offering))
            });

            return offerings;
        })
    }    
  };  